### ISSUE TITLE
Roslyn.Services.Editor.CSharp.UnitTests periodically asserts and hangs the unit tests

### ISSUE THREAD

**Issue by amcasey** [dotnet/roslyn#7825]
http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/2555/console

`[xUnit] [ERROR] - The result file 'd:\j\workspace\roslyn_prtest_win_dbg_unit64\Binaries\Debug\xUnitResults\Roslyn.Services.Editor.CSharp.UnitTests.dll.xml' for the metric 'xUnit.Net' is empty. The result file has been skipped.`

I remember seeing one yesterday as well.  I'll add a link if I can find it.

Bounty

**Comment by amcasey** [comment]
Yesterday's was in VB:

http://dotnet-ci.cloudapp.net/view/roslyn/job/roslyn_master_win_dbg_unit32/612/

**Comment by amcasey** [comment]
Yesterday's bug: https://github.com/dotnet/roslyn/issues/7807

**Comment by amcasey** [comment]
@jasonmalinowski Can you please assign this out for investigation?  We're seeing it quite frequently and we don't know which test to skip to unblock Jenkins.

**Comment by Pilchie** [comment]
@brettfo can you take a look?  Also - shouldn't we be installing something so that asserts don't put up a dialog during unit test runs?  (cc @davkean)

**Comment by davkean** [comment]
We're still pop'ing assert dialogs, we're tracking it with: https://github.com/dotnet/roslyn/issues/2796.

**Comment by davkean** [comment]
We're continually hitting this: https://github.com/dotnet/roslyn/issues/7610. Please don't close this bug without unless you've fixed it or added additional information to make diagnosing this issue easier. I've found that I can hit this assert, if I continually run the tests over and over.

**Comment by genlu** [comment]
Hit this in #8089 

http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/3050/

**Comment by davkean** [comment]
And Jason hit it too: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/3048/

@brettfo This is currently marked as "P0" - what's the status of this?

**Comment by davkean** [comment]
@brettfo I've not seen any progress on this - and we're still consistently hit this.

**Comment by brettfo** [comment]
Apologies, I've been working in the dark.  While investigating I found another flaky test and #8185 was part of dealing with that.  I'm still working on this one.

**Comment by brettfo** [comment]
As another follow up, we have a dump of the xunit process as it was hung.  I've posted a question in the xunit slack channel asking them what type would contain the name of the hanging test and until I hear back on that I'll keep poking around in WinDbg to try and find the answer.

**Comment by brettfo** [comment]
Keeping my notes public, after digging through the dump I've found that the hanging test is [UserHiddenDiagnosticSuppressionTests](http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/Diagnostics/Suppression/SuppressionTests.cs,bf3c1a49a21ba010).  I've started running that test in a loop while I explore the code to try and find the hang.

**Comment by davkean** [comment]
:+1: Thanks for keeping me the loop

**Comment by davkean** [comment]
Manish, this looks you guys own this test. This is currently, the highest flaky test that we current hit on Jenkins. Please invesigate.

**Comment by mavasani** [comment]
@brettfo What would be the easiest way to get this to repro - just run the test in a loop?

**Comment by brettfo** [comment]
That was my guess, but I haven't been able to get it to fail, even leaving the test looping over night.  I tried stepping through each call to try and find a race or anything that would cause the hang, but I couldn't find it.

FWIW, Jason has a dump of the hung xunit process.  It's on an internal share so I'll message it to you.

**Comment by mavasani** [comment]
@davkean There is not much actionable here - we haven't been able to repro the hang locally and the xunit dump is not sufficient to identify even the call stacks.

**Comment by davkean** [comment]
Please make it actionable - add logging or whatever's required to figure out the cause. This test's failure is too high and if we can't get to the bottom of it, I'm disabling it.

**Comment by mavasani** [comment]
Our investigations show that this test in isolation is not hanging. I think disabling the test is the correct step forward on this. Once you disable it, can you please open a separate issue to re-enable it and assign it to me?

**Comment by davkean** [comment]
Let's leave this issue tracking the work to fix the test, it contains all the context.

**Comment by mavasani** [comment]
@davkean Our investigations showed that this test is very likely not the cause for the hang. I had couple of questions for you:

1. Are we still seeing hangs in this unit test project?
2. Did we disable this test? If not, I would like to disable this test and see if the hang goes away or not to confirm whether this test is causing issues.

**Comment by mavasani** [comment]
@dotnet/roslyn-infrastructure Can you confirm if we are still seeing this hang? I plan to disable this test to confirm if this test is indeed the reason for the hang.

**Comment by davkean** [comment]
While we have't seen it a couple of weeks - that doesn't mean it has been fixed.

**Comment by davkean** [comment]
His this again: Roslyn.Services.Editor.CSharp.UnitTests.dll.xml

**Comment by mavasani** [comment]
@davkean Did we collect a dump? If so, can you point me to it?

**Comment by davkean** [comment]
And against: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/4629/consoleFull#-5984853535fabbf33-db14-4d24-b37d-77f64eb92e7a

**Comment by davkean** [comment]
We hit this again: https://github.com/dotnet/roslyn/issues/7825

**Comment by davkean** [comment]
And again: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/5207/consoleFull#-2115948695fabbf33-db14-4d24-b37d-77f64eb92e7a

**Comment by davkean** [comment]
And again: http://dotnet-ci.cloudapp.net/job/roslyn_future-stabilization_win_dbg_unit32/22/consoleFull#-468716515fabbf33-db14-4d24-b37d-77f64eb92e7a

**Comment by srivatsn** [comment]
Once we have Larry's watchdog process, we can get dumps for these cases to investigate if the deadlock is in xunit code.

**Comment by davkean** [comment]
This build is possible another: http://dotnet-ci.cloudapp.net/job/roslyn_stabil_win_rel_unit32/193/consoleFull#1304221255fabbf33-db14-4d24-b37d-77f64eb92e7a.

There were three test assemblies that failed to finish:
Roslyn.Services.Editor.CSharp.UnitTests.dll
Roslyn.Services.Editor.VisualBasic.UnitTests.dll
Roslyn.Compilers.VisualBasic.Emit.UnitTests.dll
Roslyn.Compilers.VisualBasic.Semantic.UnitTests.dll

Hard to tell exactly what happened.

**Comment by davkean** [comment]
@mavasani Did you look at the dump at \\<team public share>\jasonmal\Dumps\Roslyn\7610\  when you investigated this? If not, can we assign this out to figure out if this an xunit deadlock.

**Comment by davkean** [comment]
And again: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/5302/consoleFull#-2115948695fabbf33-db14-4d24-b37d-77f64eb92e7a

**Comment by mavasani** [comment]
@srivatsn did we figure out who is going to look at this one?

**Comment by tamlin-mike** [comment]
Since it seems you are currently stuck on this, I thought I'd have a quick peek at it with fresh eyes, and I have seen two areas I'd myself initially focus on:

1. Disk I/O
Differences between Jenkins and the machines where it doesn't repro.
Disk I/O delays?
In Microsoft.CodeAnalysis.Editor.Implementation.MetadataAsSource.CleanupGeneratedFiles()
I see a call:
 Directory.Delete(_rootTemporaryPath);
that is not retried in case it fails.

Perhaps there's a similar pattern for deleting files?

Having seen races when it comes to filesystem stuff, including on deleting files, I'd myself add both a log in the catch in that method, and a retry counter (with small delay between retries).


2. await/asynch
Is there any chance, however remotely, one of the await's doesn't get signaled when the asynch method ends?

Could it be feasible to add a wrapper around await/asynch dispatching for the tests, with a timeout and a unique name string? Then it could automatically log all watchdog timer failures, and pinpoint exactly when and where.

++luck;

**Comment by srivatsn** [comment]
@lgolding is going to try to get to the bottom of this.
